content/blog/2022/2022-08-02-performance-with-completablefutures.md

```diff
@@ -0,0 +1,324 @@
+interfaces. Let's see now how we can manage each of the limitations discussed above using `CompletableFuture` class.
+
+- We can complete manually the operation by providing a value
+
+ ```java
+class Demo {
+
+    public static void main(String[] args) {
+        CompletableFuture<String> stringCompletableFuture = new CompletableFuture<>();
+        stringCompletableFuture.complete("completed");
+        System.out.println("Is the stringCompletableFuture done ? " + stringCompletableFuture.isDone());
+    }
+}
+```
+
+Here we are creating a `CompletableFuture` of type `String` and we are completing it by using the method `complete()`
+passing a value. In the end, we are testing if really `stringCompletableFuture` has a value by using the
+method `isDone()` which returns `true` if completed in any fashion: normally, exceptionally, or via cancellation. The
+output of the `main()` method is:
+
+`Is the stringCompletableFuture done ? true`
+
+- We can combine one or more asynchronous operations
+
+Let's imagine that we need to call two remote APIs , `firstApiCall()` and `secondApiCall()`. The result of the first API
+will be the input for the second API. We are using the method `supplyAsync()` of the `CompletableFuture` class which
+returns a new `CompletableFuture` that is asynchronously completed by a task running in the `ForkJoinPool.commonPool()`
+with the value obtained by calling the given `Supplier`. After that we are taking the result of the `firstApiCall()` and
+using the method `thenApply()`, we pass it to the other API invoking `secondApiCall()`.
+
+```java
+class Demo {
+    public static void main(String[] args) {
+
+        var finalResult = CompletableFuture.supplyAsync(() -> firstApiCall(someValue))
+                .thenApply(firstApiResult -> secondApiCall(firstApiResult));
+    }
+}
+```
+
+- There is no way to react to the completion of the asynchronous operation without blocking the thread
+
+Using `CompletableFuture` we can now create a pipeline of asynchronous operations. Let's see a simple example of how to
+do it:
+
+```java
+class Demo {
+    public static void main(String[] args) {
+
+        CompletableFuture.supplyAsync(() -> "hello completable future")
+                .thenApply(String::toUpperCase)
+                .thenAccept(System.out::println);
+    }
+}
+```
+
+In the example above we can notice how simple is to create such a pipeline.
+
+* First, we are firstly calling the `supplyAsync()` method which takes a `Supplier` and returns a
+  new `CompletableFuture`.
+* Then, we are then transforming the result to an uppercase string by calling `thenApply()` method.
+* In the end, we just print the value on the console using `thenAccept()` that takes a `Consumer` as the argument.
+
+If we step back for a moment, we realize that working with `CompletableFuture` is very similar to Java Streams.
+
+## Performance Gains With `CompletableFuture`
+
+### Performance Using Synchronous Calls
```

Comment:
Please add an introductory section between the `##` and the `###` headers. It looks a bit lost, otherwise.

Reply:
OK

Reply:
done